### DISCUSSION TITLE
How to get route info logs from FastAPI application

### DISCUSSION THREAD

**Discussion by ashrafguitoni** [nginx/unit#1112] [Support] [answered]
When deploying a FastAPI application with uvicorn or something similar, we usually see logs such as:
`"GET / HTTP/1.1" 200 OK`

But when we switched to Nginx Unit, we can't see such logs anymore. Is there a known method for restoring this behavior?

**Comment by ac000** [comment]
Not sure if this is exactly what you mean but you can configure access logs in Unit via <https://unit.nginx.org/configuration/#access-log>

**Reply by ashrafguitoni** [reply]
Thank you for the reply @ac000, that's more or less what I want. Because I'm running on 1.26, the syntax in the doc produces this error:
```json
{
    "error": "Invalid configuration.",
    "detail": "The \"access_log\" value must be a string, but not an object."
}
```

**Reply by ashrafguitoni** [reply]
Also, sorry if I wasn't clear, but I would like to have these logs printed in STDOUT, not just a log file.

**Reply by ashrafguitoni** [reply]
We're running with `--no-daemon` and the following config (on Kubernetes):
```json
{
  "applications": {
    "fastapi": {
      "type": "python 3.9",
      "working_directory": "/app",
      "home": "/opt/venv/",
      "path": ["/app"],
      "module": "ABC.common.server",
      "callable": "APP"
    }
  },
  "listeners": {
    "*:8877": {
      "pass": "applications/fastapi"
    }
  }
}
```

**Reply by tippexs** [reply, marked as answer]
Hi -`--no-daemon` is fine. The `access_log` can be pointed to `stdout`. Your configuration will look like
```json
{
  "applications": {
    "fastapi": {
      "type": "python 3.9",
      "working_directory": "/app",
      "home": "/opt/venv/",
      "path": ["/app"],
      "module": "ABC.common.server",
      "callable": "APP"
    }
  },
  "listeners": {
    "*:8877": {
      "pass": "applications/fastapi"
    }
  },
  "access_log": "/dev/stdout"
}
```

As you are running on a quite old version (please consider to update to 1.31.1) the docs are pointing to the latest version. With 1.26 there was no configurable access log format. This was introduces in 1.28.

Update: My minimal Docker-Test:
Unit configuration for Unit 1.26.0
```json
{
  "listeners": {
    "*:8877": {
      "pass": "routes"
    }
  },
  "routes": [
	  {"action": {"return": 200}}
  ],
  "access_log": "/dev/stdout"
}
```

In Docker, this will produce the following output:

```bash
root@7d75da91cf84:/# unitd --no-daemon
2024/02/05 08:13:59 [info] 381#381 unit started
2024/02/05 08:13:59 [info] 382#382 discovery started
2024/02/05 08:13:59 [notice] 382#382 no modules matching: "/usr/lib/unit/modules/*.unit.so" found
2024/02/05 08:13:59 [info] 381#381 controller started
2024/02/05 08:13:59 [notice] 381#381 process 382 exited with code 0
2024/02/05 08:13:59 [info] 384#384 router started
2024/02/05 08:13:59 [info] 384#384 OpenSSL 1.1.1k  25 Mar 2021, 101010bf
^[[A127.0.0.1 - - [05/Feb/2024:08:16:46 +0000] "GET / HTTP/1.1" 200 0 "-" "curl/7.74.0"
127.0.0.1 - - [05/Feb/2024:08:16:49 +0000] "GET / HTTP/1.1" 200 0 "-" "curl/7.74.0"
127.0.0.1 - - [05/Feb/2024:08:16:50 +0000] "GET / HTTP/1.1" 200 0 "-" "curl/7.74.0"
127.0.0.1 - - [05/Feb/2024:08:16:51 +0000] "GET / HTTP/1.1" 200 0 "-" "curl/7.74.0"
127.0.0.1 - - [05/Feb/2024:08:16:52 +0000] "GET / HTTP/1.1" 200 0 "-" "curl/7.74.0"
```
Please note, if you would like to update to a later Unit Version please not we are now an official Docker Image. So please use `unit` instead of `nginx/unit`.

**Reply by ashrafguitoni** [reply]
Thank you for the very helpful answer. We would like to upgrade to Unit 1.30+ as well, it's just that we need to use Python 3.9 and the latest official `unit` image we found that runs on that version is 1.26. We'll consider building our own image!

**Reply by callahad** [reply]
> we need to use Python 3.9

May I ask what keeps you on 3.9? Knowing that will help us plan for our future releases of Unit and its Docker images. :)

**Reply by ashrafguitoni** [reply]
We use a lot of machine learning and scientific computation libraries in our microservices. Upgrading python versions may break or at least slightly alter how these libraries behave because they often use low-level C instructions. It may work just fine, but it may not, and that's risky for our team at the time!

**Reply by ashrafguitoni** [reply]
Also, by any chance, is it possible to filter these logs to exclude healthchecks for example?

**Comment by lcrilly** [comment]
> Also, by any chance, is it possible to filter these logs to exclude healthchecks for example?

Conditional logging was committed last week and will be in the next release. See https://github.com/nginx/unit/pull/1044

**Reply by ashrafguitoni** [reply]
Oh that's great! If you could consider releasing official images for all relatively recent python versions (e.g. 3.9, 3.10, 3.11), I think that might help us and other companies in adopting Unit. cc @callahad

**Reply by callahad** [reply]
Thank you for sharing your context. I expect us to invest more in Python support this year, and discussing what versions we support is a key consideration.
